src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -0,0 +1,142 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Connections;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Middleware;
+
+internal sealed class TlsListenerMiddleware
+{
+    private readonly ConnectionDelegate _next;
+    private readonly Action<ConnectionContext, ReadOnlySequence<byte>> _tlsClientHelloBytesCallback;
+
+    public TlsListenerMiddleware(ConnectionDelegate next, Action<ConnectionContext, ReadOnlySequence<byte>> tlsClientHelloBytesCallback)
+    {
+        _next = next;
+        _tlsClientHelloBytesCallback = tlsClientHelloBytesCallback;
+    }
+
+    /// <summary>
+    /// Sniffs the TLS Client Hello message, and invokes a callback if found.
+    /// </summary>
+    internal async Task OnTlsClientHelloAsync(ConnectionContext connection)
+    {
+        var input = connection.Transport.Input;
+
+        while (true)
+        {
+            var result = await input.ReadAsync();
+            var buffer = result.Buffer;
+
+            try
+            {
+                // If the buffer length is less than 6 bytes (handshake + version + length + client-hello byte)
+                // and no more data is coming, we can't block in a loop here because we will not get more data
+                if (result.IsCompleted && buffer.Length < 6)
+                {
+                    break;
+                }
+
+                var parseState = TryParseClientHello(buffer, out var clientHelloBytes);
+
+                if (parseState == ClientHelloParseState.NotEnoughData)
+                {
+                    // if no data will be added, and we still lack enough bytes
+                    // we can't block in a loop, so just exit
+                    if (result.IsCompleted)
+                    {
+                        break;
+                    }
+
+                    continue;
+                }
+
+                if (parseState == ClientHelloParseState.ValidTlsClientHello)
+                {
+                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
+                }
+
+                Debug.Assert(parseState is ClientHelloParseState.ValidTlsClientHello or ClientHelloParseState.NotTlsClientHello);
+                break; // We can continue with the middleware pipeline
+            }
+            finally
+            {
+                input.AdvanceTo(buffer.Start);
```

Comment:
We should be passing a value for consumed here. A bad actor could send 1 byte and then just idle and we'd be eating up a core reading and parsing the same byte over and over.

A simple standalone repro to show this:
```csharp
var p = new Pipe();
await p.Writer.WriteAsync(new byte[1]);

while (true)
{
    var res = p.Reader.ReadAsync();
    p.Reader.AdvanceTo(res.Result.Buffer.Start);
}
```

In the case where we got a full tls client hello, we'd want to set consumed to `buffer.Start` so the Https middleware will be able to read immediately.

Reply:
added a test where I write more and more bytes consequently and ensure I am reading from the buffer deterministic amount of times (5 in my test case). 

If it is an exit scenario (not valid TLS or valid TLS), then I am resetting the buffer back to `buffer.Start`: 
```
```